### PR TITLE
dbus/1.12.20: Use CMake to build D-Bus to fix cross-compiling

### PR DESCRIPTION
**Pull request by jwillikers**
Currently, Autotools errors out when attempting to cross-compile D-Bus.
However, D-Bus can be built with CMake by defining a simple variable.
Using CMake, D-Bus is able to be cross-compiled successfully.
This minimizes the complexity of the D-Bus Conanfile by using 1 build system.

D-Bus has better support for CMake in newer versions.
This should lead to reduced maintenance overhead in the future, too.

Specify library name and version:  **dbus/1.12.20**

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by SSE4 — DISMISSED** [review]

**Comment by jwillikers** [comment]
@ericLemanissier PTAL

File: recipes/dbus/1.x.x/conanfile.py [hunk 1]
```diff
@@ -103,53 +75,51 @@ def _configure_cmake(self):
...
-            cmake = self._configure_cmake()
-            cmake.build()
-        else:
-            autotools = self._configure_autotools()
-            autotools.make()
+        replace_in_file(
+            self,
+            os.path.join(self._source_subfolder, "cmake", "CMakeLists.txt"),
+            "project(dbus)",
+            "project(dbus)\ninclude(../../conanbuildinfo.cmake)\nconan_basic_setup()")
```
Comment (prince-chrismc) [inline recipes/dbus/1.x.x/conanfile.py:96]: Is there a particular reason for not using our usual `cmake-wrapper` ?

**Review by prince-chrismc — COMMENTED** [review]
One question

File: recipes/dbus/1.x.x/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/dbus/1.x.x/conanfile.py:96]: Usual `cmake-wrapper`? You mean a `CMakeLists.txt` that just includes the source one with `add_subdirectory` while doing the Conan setup? That does seem like a better idea.

Comment (jwillikers) [reply recipes/dbus/1.x.x/conanfile.py:96]: I tried. The `CMakeLists.txt` files in this version of the project are poorly written and would require patching multiple things that incorrectly rely on the path of the top-level `CMakeLists.txt` file and not the current or project-level `CMakeLists.txt` location.

Comment (jwillikers) [reply recipes/dbus/1.x.x/conanfile.py]: There were only a couple of places I had to patch. I've included the patches and the `cmake-wrapper`.

**Comment by tsondergaard** [comment]
I would also like to see this recipe simplified to only use cmake. The reason is that I have found out that the autotools built binaries include RPATHs into the build-time conan cache:

```
[ts@spark data]$ readelf -d ./dbus/1.12.20/_/_/package/af2fb75d382aa358fb89d3897162a1e6405a8f6c/bin/dbus-daemon |grep -i path 0x000000000000001d (RUNPATH)            Library runpath: [/home/ts/.conan/data/expat/2.4.8/_/_/package/fd94f5104be4181d86e876412aa89903e3d577d6/lib:/home/ts/.conan/data/dbus/1.12.20/_/_/package/af2fb75d382aa358fb89d3897162a1e6405a8f6c/lib]
```

Instead of those RPATHs it should contain just `$ORIGIN/../lib` just as e.g. the cci packages for qt and protobuf does it, so dbus-daemon and dbus-launch can always find libdbus.so...

I have managed to find the magic incantation to add `$ORIGIN/../lib`, but it is not pretty:

```diff
diff --git a/recipes/dbus/1.x.x/conanfile.py b/recipes/dbus/1.x.x/conanfile.py
index e20a3dea8..5ab0c0c78 100644
--- a/recipes/dbus/1.x.x/conanfile.py
+++ b/recipes/dbus/1.x.x/conanfile.py
@@ -89,6 +89,7 @@ class DbusConan(ConanFile):
 
             args.append("--disable-launchd")
             args.append("--disable-systemd")
+            self._autotools.link_flags.append("-Wl,-rpath,\$$ORIGIN/../lib")
 
             self._autotools.configure(args=args, configure_dir=self._source_subfolder)
         return self._autotools
```

However, I cannot for the life of me figure out how to make libtool stop hardcoding those other library paths, so I end up with this:

```
readelf -d ~/.conan/data/dbus/1.12.20/ts/rpath-fix/package/e08832a0b34a4af568fda98c13988c9115f223f1/bin/dbus-daemon |grep -i path
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN/../lib:/home/ts/.conan/data/expat/2.4.8/_/_/package/fd94f5104be4181d86e876412aa89903e3d577d6/lib:/home/ts/.conan/data/dbus/1.12.20/ts/rpath-fix/package/e08832a0b34a4af568fda98c13988c9115f223f1/lib]
```

With CMake this would be a simple matter of setting `CMAKE_INSTALL_RPATH` to  `"$ORIGIN/../lib` to add the desired RPATH.

**Comment by tsondergaard** [comment]
@jwillikers, this PR essentially reverts the decision made in #4114 by @ericLemanissier for the same version of dbus. That weakens your statement that "D-Bus has better support for CMake in newer versions". @ericLemanissier  made #4114 to fix the build on BSD and because of the recommendation given [here](https://gitlab.freedesktop.org/dbus/dbus/-/issues/324#note_746339). Does your PR address the BSD build problem? My opinion is that it would be better to apply the necessary patches to fix the build with CMake on BSD than to use multiple build systems in the recipe. Especially as the authors also say that patches (to fix the cmake build on BSD) are welcome.

**Comment by jwillikers** [comment]
@tsondergaard Thanks for the feedback and relevant background of which I was unaware. I'll be sure to test these changes out on FreeBSD, though I'm sure as they are now they probably won't work. I agree that patching CMake to work on FreeBSD is the way to go for this, and I can surmise that this will probably be a simple fix.

I'll see if I can get these patches merged in upstream, too. That would really help the situation.

**Comment by jwillikers** [comment]
Ironically enough, looks like D-Bus is moving to Meson, which aligns it with the rest of the freedesktop ecosystem. See [here](https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/NEWS#L13-24).

**Comment by SSE4** [comment]
> Ironically enough, looks like D-Bus is moving to Meson, which aligns it with the rest of the freedesktop ecosystem. See [here](https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/NEWS#L13-24).

if it's switched to meson, please create a pull request switching conan recipe to use meson as well. it's pointless to migrate from cmake to autotools and vice versa in that case.

**Comment by jwillikers** [comment]
> > Ironically enough, looks like D-Bus is moving to Meson, which aligns it with the rest of the freedesktop ecosystem. See [here](https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/NEWS#L13-24).
> 
> if it's switched to meson, please create a pull request switching conan recipe to use meson as well. it's pointless to migrate from cmake to autotools and vice versa in that case.

I will certainly do that, but right now we'll have to wait for version 1.15 to be released before it's available.

**Comment by jwillikers** [comment]
@tsondergaard I figured the issue when building on FreeBSD using CMake. In fact, there's even an existing commit upstream for the issue. I've pulled that commit as a patch here and it now builds for me on FreeBSD using CMake.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**